### ISSUE TITLE
RegionCalculus uses a different dataset shape convention than hdf5 file

### ISSUE THREAD

**Issue by Yurlungur** [eschnett/SimulationIO#24]
At least in the release-0.1 version I'm using, the following Python code gives one answer:
```python
import numpy as np # numpy as a wrapper for fortran arrays
dataset = discretefieldblockcomponent.getData_DataSet()
region = block.region # remember our block from earlier?
# includes ghost zones
data = np.array(dataset.read_double()).reshape(region.shape())
print data.shape
```
and the following code a different answer:
```python
path = discretefieldblockcomponent.getPath()
name = discretefieldblockcomponent.getName()
import h5py
with h5py.File('static-tov-cell.s5','r') as f:
    data_h5py = f[path][name][:]
print data_h5py.shape
```

Ideally, they should give the same answer should they not?

**Comment by eschnett** [comment]
Jonah

Yes, these two should be the same.

There are two different conventions for ordering array indices, Fortran
order and C order. Julia uses Fortran order, Python uses C order. You are
probably missing a "reversed()" of the result of "region.shape()" before
calling "reshape". (I think there is also a way to tell Python to use
Fortran order.)

An array that is indexed a(i,j,k) in Fortran is indexed a[k][j][i] in C, or
a[k,j,i] in Python, and a[i,j,k] in Julia -- note the difference in the
order of the indices.

-erik



On Thu, Jun 9, 2016 at 4:35 PM, Jonah Miller <notifications@github.com>
wrote:

> At least in the release-0.1 version I'm using, the following Python code
> gives one answer:
>
> import numpy as np # numpy as a wrapper for fortran arrays
> dataset = discretefieldblockcomponent.getData_DataSet()
> region = block.region # remember our block from earlier?# includes ghost zones
> data = np.array(dataset.read_double()).reshape(region.shape())print data.shape
>
> and the following code a different answer:
>
> path = discretefieldblockcomponent.getPath()
> name = discretefieldblockcomponent.getName()import h5pywith h5py.File('static-tov-cell.s5','r') as f:
>     data_h5py = f[path][name][:]print data_h5py.shape
>
> Ideally, they should give the same answer should they not?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eschnett/SimulationIO/issues/24>, or mute the thread
> <https://github.com/notifications/unsubscribe/AANCCpdCWA_OgmIuHt6YdrYvIDyUvDKTks5qKHkAgaJpZM4IyYIC>
> .
>



-- 
Erik Schnetter <schnetter@gmail.com>
http://www.perimeterinstitute.ca/personal/eschnetter/

**Comment by Yurlungur** [comment]
Erik,

There is a way to tell Python to use Fortran order, yes. But that doesn't change the shape of the array, which is a view data structure, only how the the shape is interpreted in memory. I suppose the cleanest thing is to add a "reverse" everywhere or to tell h5py to use Fortran ordering. Surely there must be a way to automate this, though?

I take it Carpet uses Fortran ordering? 

Best,
Jonah

**Comment by eschnett** [comment]
Yes, Cactus uses Fortran ordering as well. Note that this is mostly about
the order in which the indices are used; it's never necessary to actually
transpose an array.

-erik

On Thursday, June 9, 2016, Jonah Miller <notifications@github.com> wrote:
> Erik,
>
> There is a way to tell Python to use Fortran order, yes. But that doesn't
change the shape of the array, which is a view data structure, only how the
the shape is interpreted in memory. I suppose the cleanest thing is to add
a "reverse" everywhere or to tell h5py to use Fortran ordering. Surely
there must be a way to automate this, though?
>
> I take it Carpet uses Fortran ordering?
>
> Best,
> Jonah
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.<
https://ci3.googleusercontent.com/proxy/WSSkDS-QjP4_53xWIEnljaworFLN8rSJ6-tWUSSH98kGmtRHlgUDfuTU9vSt6f_aRRpDtQvHLU-7y_Ff7jDlwK9Mx6GDJXEkxCzCxKuVDB_2jSI5_mtx4nyOflQJqcHxWypMadILUIuCfakKgF-0Qqwyd78uNQ=s0-d-e1-ft#https://github.com/notifications/beacon/AANCCkqJLeBClpTGAqiSYetMIG-z-DK2ks5qKJSDgaJpZM4IyYIC.gif
>

-- 
Erik Schnetter <schnetter@gmail.com>
http://www.perimeterinstitute.ca/personal/eschnetter/

**Comment by Yurlungur** [comment]
This is low priority, but at some point, I'd like to modify the python interface so that things like "region.shape()" return the same shape and ordering as h5py, i.e., row-major. So for that reason, I'd like to leave this issue open as a "to-do" for myself. Is that okay? Or should I put it in my repository?

**Comment by eschnett** [comment]
Changing the order probably makes sense for Python. Sure, leave it open.
